### PR TITLE
DIP 13: Identities in Hierarchical Deterministic Wallets

### PR DESCRIPTION
**Pull request by thephez**
Introduces DashPay-related DIP 13: _Identities in Hierarchical Deterministic Wallets_.

There are four DIPs related to DashPay:
- **13: Identities in Hierarchical Deterministic Wallets (this PR)**
- [14: Extended Key Derivation using 256-bit Unsigned Integers](https://github.com/dashpay/dips/pull/71)
- [15: DashPay](https://github.com/dashpay/dips/pull/72)
- [16: Headers First Synchronization on Simple Payment Verification Wallets](https://github.com/dashpay/dips/pull/73)

The original draft PR (#69) has been split into four separate PRs so each document is in its own PR.

### CODE REVIEW

**Review by QuantumExplorer — APPROVED** [review]

**Review by PastaPastaPasta — APPROVED** [review]
test approval

**Review by HashEngineering — DISMISSED** [review]

**Review by bcarrollaz — APPROVED** [review]
utACK

File: README.md [hunk 1]
```diff
@@ -30,9 +29,9 @@ Number | Layer | Title | Owner | Type | Status
...
 [11](dip-0011.md) | Consensus | Identities | Ivan Shumkov, Anton Suprunchuk, Samuel Westrich, Cofresi | Standard | Proposed
 [12](dip-0012.md) | Consensus | Dash Platform Name Service | Ivan Shumkov, Anton Suprunchuk | Standard | Proposed
+[13](dip-0013.md) | Applications | Identities in Hierarchical Deterministic Wallets | Samuel Westrich | Informational | Proposed
 [20](dip-0020.md) | Consensus | Dash Opcode Updates | Mart Mangus | Standard | Proposed
 
-
 ## License
 
-Unless otherwise specified, Dash Improvement Proposals (DIPs) are released under the terms of the MIT license. See [LICENSE](LICENSE) for more information or see https://opensource.org/licenses/MIT.
+Unless otherwise specified, Dash Improvement Proposals (DIPs) are released under the terms of the MIT license. See [LICENSE](LICENSE) for more information or see the [MIT License](https://opensource.org/licenses/MIT).
```
Comment (PastaPastaPasta) [inline README.md:37]: Generally, I'd say this should be in it's own PR, but it's probably okay :D

**Review by PastaPastaPasta — APPROVED** [review]
ACK.

Read through the document, all items make sense, this is a good thing to standardize. :+1:

File: README.md (hunk 1 above)
Comment (thephez) [reply README.md:37]: Whew :relieved:

File: dip-0013.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+derived as ECDSA keys.
+
+We define the following levels for them:
+
+`m / purpose' / coin_type' / feature' / sub feature' / funding index /`
+
+Sub feature is set to `2'`.
+
+We should note that unlike Identity Registration keys, Top Up Funding keys do not reference an
+identity and can be used to top up any identity. It is useful to point out that a registration
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply dip-0013.md:146]: I'm not sure I get your concern. Is it that you could top up identity Bob for example and then with the change of that topUp then topUp identity Alice? And then you could surmise that the identities are linked?

File: dip-0013.md [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+
+If a wallet recovers an unused top up funding transaction, the wallet should attempt to reuse it.
+This is done in a few steps. First, the wallet must recognise the key used in a returned identity
+funding transaction as a top up funding key. Then the wallet must query Dash Platform to check if
+the top up transaction has already been used and the balance transferred into credits. If this is
+not the case, the wallet can then perform an identity top up transition as defined in DIP11 using
+the top up funding key.
+
+It is recommended that 30 address hashes of these keys be added to the bloom filter sent to peers.
+If 25 identities have been found, the filter should be regenerated with 30 more.
```
[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply dip-0013.md:158]: This has been a while since I've worked on this part so I hope my memory is good enough. Since the bloom filter is made for the peer in question and since clients switch peers when changing bloom filters (to maintain privacy) it is essentially regenerated.

[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply dip-0013.md:158]: And you found an error @UdjinM6 ! it's not 25 identities, it's 25 addresses. I'll fix this.

File: dip-0013.md [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+must use HASH160 of the public key which is one round of SHA256 followed by one round of RIPEMD160.
+
+# Identity Top Up Funding keys
+
+Identity Funding keys used for topup are the third sub-feature for identities. They are always
+derived as ECDSA keys.
+
+We define the following levels for them:
+
+`m / purpose' / coin_type' / feature' / sub feature' / funding index /`
```
[the remark this answers is not in the document]
Comment (thephez) [reply dip-0013.md:141]: @QuantumExplorer Do you concur with the suggestion?

**Review by QuantumExplorer — DISMISSED** [review]

File: dip-0013.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (UdjinM6) [reply dip-0013.md:146]: > Is it that you could top up identity Bob for example and then with the change of that topUp then topUp identity Alice?

Yes, smth like that. Also: 1. can you topup someone else's identity? 2. will the change be sent back to the same address (and cause all the known address reuse issues) or how is this change address/key generated/managed? If you topup one identity and the change is then used to topup another identity and the change from this tx is later used like that again but you can topup your own identities only, one can assume both of these are yours. If you can topup identities of other people, one can also additionally assume that there might be some link between you and another person (friends, family, colleges etc.). So basically, whenever there is a change in multi-identity scenario you leak some info. Why don't we simply tie topup funding keys to identities like so `m / purpose' / coin_type' / feature' / sub feature' / identity index / funding index` (0 - registration, 1+ - topups)?

[the remark this answers is not in the document]
Comment (thephez) [reply dip-0013.md:146]: > > Is it that you could top up identity Bob for example and then with the change of that topUp then topUp identity Alice?
> 
> Yes, smth like that. Also: 1. can you topup someone else's identity? 

Re: 1 - Yes, anyone can topup any identity regardless of if it is theirs or not. For example, App devs could provide topups to cover their user's fees, etc.

[the remark this answers is not in the document]
Comment (thephez) [reply dip-0013.md:146]: @QuantumExplorer Can you take a look at Udjin's response?

[the remark this answers is not in the document]
Comment (QuantumExplorer) [reply dip-0013.md:146]: @UdjinM6 I agree that there is not much privacy but privacy was not so much the focus of this DIP. This dip is really about derivation paths in HD wallets for making keys. Having an HD wallet or not does not really make a difference in terms of privacy unless I am missing something. I would definitely think that we could and probably should investigate how to make things more private, but I think that should be a different dip and would be happy to collaborate with you to figure it out if you want.

**Review by UdjinM6 — APPROVED** [review]

**Review by QuantumExplorer — APPROVED** [review]
